### ISSUE TITLE
Add a "patch" source type

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#63]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/63)
In GitLab by [[Gitlab user @bochecha]](https://gitlab.com/bochecha) on Aug 7, 2017, 15:25

It's currently possible to add patches as a `local` source: https://gitlab.com/BuildStream/gnome-test-data/blob/48e30c94a85477566c5c77d2790545955d12c6c2/elements/core-deps/mozjs38.bst

It would be much nicer to be able to add `patch` sources, for example something like this:

```yaml
sources:
- kind: tar
  url: ...
  ref: ...
- kind: patch
  path: ...
  strip-components: 0
```

The `strip-components` would default to `1` (like in `flatpak-builder` manifests) as that's probably the most common these days given how everybody uses Git.

BuildStream would then automatically patch the sources with the `patch` sources, in the order they are declared.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Aug 21, 2017, 20:17

This sounds like a sensible convenience; it could be a `Source` plugin which applies the patch at `Source.stage()` time and raises an error in case of staging into an empty directory.

Would have to raise an appropriate error if the user tries to open a workspace on a `patch` source. Handling patch workspaces in a custom way is probably not worthwhile; especially since this one would probably only support local patches (not remote URIs).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Sep 7, 2017, 13:40

mentioned in merge request !86

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 9, 2017, 23:11

closed via merge request !86

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Sep 9, 2017, 23:11

closed via commit b1b5cd56386272b04487eb7563198215ce9ae9a1
